File: packages/bridge/src/stt/sidecar.py
"""Conn STT sidecar — push-to-talk capture + local transcription.

Spawned once by the bridge. Owns the microphone (sounddevice) and a
faster-whisper model (CPU int8). Protocol: one JSON object per line.

  stdin:  {"cmd": "ping"}    -> {"ok": true}
          {"cmd": "start"}   -> {"ok": true}            (begin capture)
          {"cmd": "stop"}    -> {"text": "...", "ms": 812}
          {"cmd": "cancel"}  -> {"ok": true}            (discard capture)
  stdout: protocol lines only; {"event": "ready"} once the model is loaded.
          Logs go to stderr (the bridge forwards them to its log).

The model download (first run only) happens during load; the bridge shows
the key as "loading" until the ready event.
"""

import io
import json
import sys
import time
import wave
import argparse
import threading

SAMPLE_RATE = 16000


def out(obj):
    sys.stdout.write(json.dumps(obj) + "\n")
    sys.stdout.flush()


def log(msg):
    sys.stderr.write(f"[stt] {msg}\n")
    sys.stderr.flush()


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--model", default="distil-small.en")
    ap.add_argument("--language", default="en")
    ap.add_argument("--device", default=None, help="sounddevice input device name/index")
    args = ap.parse_args()

    # Import late so a missing dependency produces a clean stderr line the
    # bridge can log, instead of a stack trace before stdio is wired.
    try:
        import sounddevice as sd
        from faster_whisper import WhisperModel
    except Exception as e:  # noqa: BLE001
        log(f"dependency import failed: {e}")
        sys.exit(2)

    log(f"loading model {args.model} (cpu int8)…")
    t0 = time.time()
    model = WhisperModel(args.model, device="cpu", compute_type="int8")
    log(f"model loaded in {time.time() - t0:.1f}s")

    chunks: list[bytes] = []
    recording = threading.Event()
    lock = threading.Lock()

    def on_audio(indata, frames, t, status):  # noqa: ANN001
        if status:
            log(f"audio status: {status}")
        if recording.is_set():
            with lock:
                chunks.append(bytes(indata))

    device = None
    if args.device is not None:
        device = int(args.device) if args.device.isdigit() else args.device

    # The microphone is opened ON DEMAND, not at startup. Holding an input
    # stream open for the whole session lights the OS "in use" indicator and
    # claims the headset all day for a feature used in bursts — on a machine
    # whose owner is frequently on calls, that is not a reasonable default.
    # The MODEL stays loaded (that's the ~4s cost); opening a device is not.
    stream = None

    def open_mic() -> int:
        """Open + start the input stream. Returns ms taken."""
        nonlocal stream
        if stream is not None:
            return 0
        t0 = time.time()
        s = sd.RawInputStream(
            samplerate=SAMPLE_RATE,
            channels=1,
            dtype="int16",
            device=device,
            callback=on_audio,
        )
        s.start()
        stream = s
        ms = int((time.time() - t0) * 1000)
        log(f"microphone open: {sd.query_devices(s.device)['name']} ({ms}ms)")
        return ms

    def close_mic() -> None:
        nonlocal stream
        if stream is None:
            return
        s, stream = stream, None
        try:
            s.stop()
            s.close()
        except Exception as e:  # noqa: BLE001
            log(f"microphone close failed: {e}")
        else:
            log("microphone released")

    out({"event": "ready", "model": args.model})

    def transcribe(pcm: bytes) -> str:
        if len(pcm) < SAMPLE_RATE // 5 * 2:  # <200ms of audio — nothing to say
            return ""
        bio = io.BytesIO()
        with wave.open(bio, "wb") as w:
            w.setnchannels(1)
            w.setsampwidth(2)
            w.setframerate(SAMPLE_RATE)
            w.writeframes(pcm)
        bio.seek(0)
        segments, _info = model.transcribe(
            bio,
            language=args.language,
            beam_size=1,
            vad_filter=True,
            condition_on_previous_text=False,
        )
        return " ".join(s.text.strip() for s in segments).strip()

    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        try:
            cmd = json.loads(line).get("cmd")
        except json.JSONDecodeError:
            out({"error": "bad json"})
            continue
        if cmd == "ping":
            out({"ok": True})
        elif cmd == "start":
            with lock:
                chunks.clear()
            # Open BEFORE replying: the deck only shows REC once this returns,
            # so the key can't invite speech into a device that isn't live yet.
            try:
                open_ms = open_mic()
            except Exception as e:  # noqa: BLE001
                log(f"microphone open failed: {e}")
                out({"ok": False, "error": "mic unavailable"})
                continue
            recording.set()
            out({"ok": True, "openMs": open_ms})
        elif cmd == "cancel":
            recording.clear()
            close_mic()
            with lock:
                chunks.clear()
            out({"ok": True})
        elif cmd == "stop":
            recording.clear()
            # Release the device before transcribing — transcription takes far
            # longer than recording, and nothing needs the mic during it.
            close_mic()
            with lock:
                pcm = b"".join(chunks)
                chunks.clear()
            t0 = time.time()
            try:
                text = transcribe(pcm)
            except Exception as e:  # noqa: BLE001
                log(f"transcription failed: {e}")
                text = ""
            out({"text": text, "ms": int((time.time() - t0) * 1000)})
        else:
            out({"error": f"unknown cmd: {cmd}"})


if __name__ == "__main__":
    main()

File: packages/bridge/src/stt/sidecar.ts
import { spawn, type ChildProcessWithoutNullStreams } from "node:child_process";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { createInterface } from "node:readline";
import type { Logger } from "../log.js";

/**
 * Push-to-talk sidecar adapter — one persistent Python process (sidecar.py)
 * owning the microphone and a local faster-whisper model. JSON lines over
 * stdio, one response per command, mirroring the AHK daemon pattern.
 *
 * Status drives the mic key face:
 *  - "offline": not running (deps missing, crash, disabled) — key grayed.
 *  - "loading": spawned, model loading (first run downloads it).
 *  - "ready" / "recording" / "transcribing": the live dictation cycle.
 * Failure posture per the design doc: sidecar down → offline key; audio is
 * never queued silently.
 */
export type SttStatus = "offline" | "loading" | "ready" | "recording" | "transcribing";

const SIDECAR_PATH = join(dirname(fileURLToPath(import.meta.url)), "sidecar.py");
// stop → transcription of up to maxSeconds of audio on CPU; generous.
const STOP_TIMEOUT_MS = 30_000;
const CMD_TIMEOUT_MS = 5_000;

export interface SttConfig {
  python: string;
  model: string;
  language: string;
  /** sounddevice input device name/index; omit for the system default. */
  device?: string;
}

/** Interface the controller depends on — tests stub this. */
export interface SttEngine {
  readonly status: SttStatus;
  /** Optional respawn hook: a mic press while offline retries the sidecar. */
  ensureStarted?(): Promise<void>;
  start(): Promise<boolean>;
  stop(): Promise<string>;
  cancel(): Promise<void>;
}

export class SttSidecar implements SttEngine {
  private proc: ChildProcessWithoutNullStreams | null = null;
  private queue: Promise<unknown> = Promise.resolve();
  private _status: SttStatus = "offline";
  private spawning = false;

  constructor(
    private readonly cfg: SttConfig,
    private readonly log: Logger,
    private readonly onStatus: (status: SttStatus) => void,
  ) {}

  get status(): SttStatus {
    return this._status;
  }

  private setStatus(s: SttStatus): void {
    if (this._status === s) return;
    this._status = s;
    this.onStatus(s);
  }

  /** Spawn the sidecar; resolves once ready (or offline on failure). Safe to
   * call repeatedly — a mic press while offline retries the spawn. */
  async ensureStarted(): Promise<void> {
    if (this.proc || this.spawning) return;
    this.spawning = true;
    this.setStatus("loading");
    try {
      const args = [SIDECAR_PATH, "--model", this.cfg.model, "--language", this.cfg.language];
      if (this.cfg.device) args.push("--device", this.cfg.device);
      const proc = spawn(this.cfg.python, args, { stdio: "pipe", windowsHide: true });
      this.proc = proc;
      proc.stderr.setEncoding("utf8");
      createInterface({ input: proc.stderr }).on("line", (l) => this.log.info({ stt: l }, "stt sidecar"));
      proc.on("exit", (code) => {
        this.log.warn({ code }, "stt sidecar exited");
        this.proc = null;
        this.setStatus("offline");
      });
      const lines = createInterface({ input: proc.stdout });
      // First protocol line must be the ready event; the model load (and its
      // one-time download) can take a while — no timeout here, the key just
      // shows "loading".
      const ready = await new Promise<boolean>((resolve) => {
        lines.once("line", (l) => {
          try {
            resolve(JSON.parse(l)?.event === "ready");
          } catch {
            resolve(false);
          }
        });
        proc.once("exit", () => resolve(false));
      });
      if (!ready) {
        proc.kill();
        this.proc = null;
        this.setStatus("offline");
        return;
      }
      this.lines = lines;
      this.setStatus("ready");
      this.log.info({ model: this.cfg.model }, "stt sidecar ready");
    } catch (err) {
      this.log.warn({ err: String(err) }, "stt sidecar spawn failed");
      this.proc = null;
      this.setStatus("offline");
    } finally {
      this.spawning = false;
    }
  }

  private lines: ReturnType<typeof createInterface> | null = null;

  async start(): Promise<boolean> {
    if (this._status !== "ready") return false;
    const reply = await this.command({ cmd: "start" }, CMD_TIMEOUT_MS);
    if (reply?.ok === true) {
      this.setStatus("recording");
      return true;
    }
    return false;
  }

  async stop(): Promise<string> {
    if (this._status !== "recording") return "";
    this.setStatus("transcribing");
    const reply = await this.command({ cmd: "stop" }, STOP_TIMEOUT_MS);
    // Sidecar died mid-transcription → status already "offline"; don't lie.
    if (this.proc) this.setStatus("ready");
    return typeof reply?.text === "string" ? reply.text : "";
  }

  async cancel(): Promise<void> {
    if (this._status !== "recording") return;
    await this.command({ cmd: "cancel" }, CMD_TIMEOUT_MS);
    if (this.proc) this.setStatus("ready");
  }

  dispose(): void {
    this.proc?.kill();
    this.proc = null;
    this.setStatus("offline");
  }

  /** Serialized request/response — the sidecar answers strictly in order. */
  private command(req: object, timeoutMs: number): Promise<Record<string, unknown> | null> {
    const exec = async (): Promise<Record<string, unknown> | null> => {
      if (!this.proc || !this.lines) return null;
      return new Promise((resolve) => {
        const timer = setTimeout(() => {
          this.lines?.off("line", onLine);
          resolve(null);
        }, timeoutMs);
        const onLine = (line: string) => {
          clearTimeout(timer);
          try {
            resolve(JSON.parse(line));
          } catch {
            resolve(null);
          }
        };
        this.lines!.once("line", onLine);
        this.proc!.stdin.write(JSON.stringify(req) + "\n");
      });
    };
    const next = this.queue.then(exec, exec);
    this.queue = next.catch(() => {});
    return next as Promise<Record<string, unknown> | null>;
  }
}
